content/blog/prototyping/index.md

```diff
@@ -0,0 +1,726 @@
+mentioned that premature optimization is one of the worst things you can do in Rust. 
+Especially experienced developers coming from C or C++ might be tempted to optimize too early.
+While prototyping, this can obfuscate simpler designs and make refactoring harder later on.
+Resist the urge to optimize too early.
+You will thank yourself later. [^1]
+
+[^1]: In the talk, I show an example where early over-optimization led to the wrong abstraction and made the code slower. The actual bottleneck was elsewhere and hard to uncover without profiling.
+
+### Use `println!` and `dbg!` for debugging 
+
+I find that printing values is pretty handy while prototyping.
+It's one less context switch to make compared to starting a debugger.
+
+Most people use `println!` for that, but [`dbg!`](https://doc.rust-lang.org/std/macro.dbg.html) has a few advantages:
+
+- It prints the file name and line number where the macro is called.
+- It outputs the expression as well as its value.
+- It's less syntax-heavy than `println!`; e.g. `dbg!(x)` vs. `println!("{x:?}")`.
+- It's only active in debug builds, so you don't have to worry about removing it later.
+
+Where `dbg!` really shines is in recursive functions or when you want to see the intermediate values during an iteration:
+
+```rust
+fn factorial(n: u32) -> u32 {
+    // `dbg!` returns the argument, 
+    // so you can use it in the middle of an expression
+    if dbg!(n <= 1) {
+        dbg!(1)
+    } else {
+        dbg!(n * factorial(n - 1))
+    }
+}
+
+dbg!(factorial(4));
+```
+
+The output is nice and tidy:
+
+```rust
+[src/main.rs:2:8] n <= 1 = false
+[src/main.rs:2:8] n <= 1 = false
+[src/main.rs:2:8] n <= 1 = false
+[src/main.rs:2:8] n <= 1 = true
+[src/main.rs:3:9] 1 = 1
+[src/main.rs:7:9] n * factorial(n - 1) = 2
+[src/main.rs:7:9] n * factorial(n - 1) = 6
+[src/main.rs:7:9] n * factorial(n - 1) = 24
+[src/main.rs:9:1] factorial(4) = 24
+```
+
+If you're interested, here are [more details on how to use that handy `dbg!` macro](https://edgarluque.com/blog/rust-dbg-macro/).
+
+### Play With Types
+
+Quite frankly, the type system is one of the main reasons I love Rust.
+It feels great to express my ideas in types and see them come to life.
+I would encourage you to heavily lean into the type system during the prototyping phase.
+
+Try to build up a vocabulary of concepts and own types which describe your system. 
+In the beginning, you won't have a good idea of the types in your system.
+That's fine!
+Start with *something* and quickly sketch out solutions and gradually add constraints to model the business requirements.
+Don't stop until you find a version that feels just right. [^2]
+
+[^2]: I usually know when I found a good abstraction once I can use all of Rust's features like expression-oriented programming and pattern matching together with my own types. The types feel like a natural extension of the language and fit in seamlessly.
+
+Let's look at an example:
```

Comment:
Maybe it would help if I explain why going from `String` to an own type can help prevent errors. On the flip side, coming up with a good set of types is not easy -- and maybe that's the entire point: you need to play with types until you find a combination of types that work well together. So the example might confuse people and be beside the point I'm trying to make.

I could see myself writing a long paragraph about types alone here, and that would be highly distracting. If we can come up with a simpler example, which emphasizes my point while keeping it short and focused, that would help. Open for ideas.

Reply:
How about now? Tried my best to come up with a better example.